keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -514,310 +481,126 @@ spec:
-  # to more users than the original CRD. Both objects are in the same
-  # namespace.
-  kind: CardParameters
-  apiVersion: dra.example.com/v1alpha1
-  metadata:
-    name: my-parameters
-    namespace: user-namespace
-    uid: foobar-uid
-  ...
-  spec:
-    minimumRuntimeVersion: v12.0.0
-    minimumMemory: 32Gi
-    sharing:
-      strategy: TimeSliced
-
-driverRequests:
-- driverName: cards.dra.example.com
-  requests:
-  # Each entry here is a request for one resource.
-  - namedresources:
-      # Selectors are CEL expressions with access to the attributes of the named resource
-      # that is being checked for a match.
-      selector: |-
-        attributes.version["runtimeVersion"].isGreaterThan(semver("12.0.0")) &&
-        attributes.quantity["memory"].isGreaterThan(quantity("32Gi"))
-```
-
-The meaning is that the selector expression must evaluate to true for a
-particular named resource in `namedResources`.
-
-Future extensions could be added to support partioning of resources as well as a
-express constraints that must be satisfied *between* any selected resources. For
-example, selecting two cards which are on the same PCI root complex may be
-needed to get the required performance.
-
-Instead of defining a vendor-specific CRD, DRA driver authors (or
-administrators) could decide to allow users to create and reference
-`ResourceClaimParameters` directly within their `ResourceClaims`. This would
-avoid the translation step shown above, but at the cost of (1) providing per-
-claim configuration parameters for their requested resources, and (2) doing any
-sort of validation on the CEL expressions created by the user.
-
-Resource class parameters are supported the same way. To ensure that
-permissions can be limited to administrators, there's a separate cluster-scoped
-ResourceClassParameters type. Instead of individual requests, one additional
-selector can be specified there which then also must be true for all individual
-requests made with that class:
-
-```yaml
-kind: ResourceClassParameters
-apiVersion: resource.k8s.io/v1alpha2
-
-metadata:
-  name: someArbitraryName
-
-generatedFrom:
-  name: gpu-parameters
-  kind: CardClassParameters
-  apiGroup: dra.example.com
-  uid: foobar-uid
-
-vendorParameters:
-  ...
-
-filters:
-- driverName: cards.dra.example.com
-  namedResources:
-    selector: |-
-      attributes["memory"] <= "16Gi"
-```
-
-In this example, the additional selector expression limits users of this class
-to just the cards with less that "16Gi" of memory. Together with limiting the
-number of claims that users are allowed to create for this class (see resource
-quotas in the core KEP) this can ensure that users do not consume too many
-resources. Allowing resource quotas that are based on resource attributes may be
-a useful future enhancement.
+A DeviceClass can contain the same device selector and device configuration
+parameters. Those get added to what is specified in the claim when a class gets
+referenced.
```

Comment:
Which takes precedence when a the configuration is a class and a claim conflict?

Reply:
AIUI, both get passed to the driver, which makes the decision. Patrick, can the driver tell which came from class and which from claim?

Reply:
Yes. There is `DeviceConfiguration.Admin` for that.